File: jscomp/others/bs_HashMapInt.ml
# 9 "hashmap.cppo.ml"
type key = int
type seed = int
external caml_hash_mix_int : seed -> int -> seed  = "caml_hash_mix_int"
external final_mix : seed -> seed = "caml_hash_final_mix"
let hash (s : key) = 
  final_mix (caml_hash_mix_int 0 s)

# 19
(***********************************************************************)
(*                                                                     *)
(*                                OCaml                                *)
(*                                                                     *)
(*            Xavier Leroy, projet Cristal, INRIA Rocquencourt         *)
(*                                                                     *)
(*  Copyright 1996 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  under the terms of the GNU Library General Public License, with    *)
(*  the special exception on linking described in file ../LICENSE.     *)
(*                                                                     *)
(***********************************************************************)
(**  Adapted by Authors of BuckleScript 2017                           *)
type ('a, 'b,'id) t0 = ('a,'b,'id) Bs_internalBuckets.t0 =
  {
    mutable size: int;                        (* number of entries *)
    mutable buckets: ('a, 'b) Bs_internalBuckets.bucketlist array;  (* the buckets *)
    initial_size: int;                        (* initial array size *)
  }


type ('a,'b) buckets = ('a,'b) Bs_internalBuckets.buckets =
  {
    mutable key : 'a ; 
    mutable value : 'b ; 
    mutable next : ('a, 'b) buckets Bs_internalBuckets.opt
  }

type 'b t = (key,'b,unit) t0


let rec insert_bucket  ~h_buckets ~ndata_tail h old_bucket = 
  match Bs_internalBuckets.toOpt old_bucket with 
  | None -> ()
  | Some ({key; next} as cell) ->
    (* let nidx = key_index ~hash h key in *)
    let nidx = hash key land (Array.length h_buckets - 1) in 
    let v = Bs_internalBuckets.return cell in 
    begin match Bs_internalBuckets.toOpt (Bs_Array.unsafe_get ndata_tail nidx) with
      | None -> 
        Bs_Array.unsafe_set h_buckets nidx  v
      | Some tail ->
        tail.next <- v ; (* cell put at the end *)            
    end;          
    Bs_Array.unsafe_set ndata_tail nidx  v;
    insert_bucket  ~h_buckets ~ndata_tail h next


let resize  h =
  let odata = h.buckets in
  let osize = Array.length odata in
  let nsize = osize * 2 in
  if nsize >= osize then begin (* no overflow *)
    let h_buckets = Bs_internalBuckets.makeSize nsize  in
    let ndata_tail = Bs_internalBuckets.makeSize nsize  in (* keep track of tail *)
    h.buckets <- h_buckets;          (* so that indexfun sees the new bucket count *)
    for i = 0 to osize - 1 do
      insert_bucket  ~h_buckets ~ndata_tail h (Bs_Array.unsafe_get odata i)
    done;
    for i = 0 to nsize - 1 do
      match Bs_internalBuckets.toOpt (Bs_Array.unsafe_get ndata_tail i) with
      | None -> ()
      | Some tail -> tail.next <- Bs_internalBuckets.emptyOpt
    done
  end


let add  h key value =
  let h_buckets = h.buckets in  
  let h_buckets_lenth = Array.length h_buckets in 
  let i =  hash key land (h_buckets_lenth - 1) in 
  let bucket = {key; value; next= Bs_Array.unsafe_get h_buckets i} in  
  Bs_Array.unsafe_set h_buckets i  (Bs_internalBuckets.return bucket);
  let h_new_size = h.size + 1 in 
  h.size <- h_new_size;
  if h_new_size > h_buckets_lenth lsl 1 then resize  h


let rec remove_bucket h h_buckets  i key prec buckets =
  match Bs_internalBuckets.toOpt buckets with
  | None -> ()
  | Some {key=k; next} ->
    if  (k : key) = key 
    then 
      begin
        (match Bs_internalBuckets.toOpt prec with
         | None -> Bs_Array.unsafe_set h_buckets i  next
         | Some c -> c.next <- next);
        h.size <- h.size - 1;        
      end
    else remove_bucket  h h_buckets i key buckets next

let remove  h key =  
  let h_buckets = h.buckets in 
  (* let i = key_index ~hash h key in *)
  let i = hash key land (Array.length h_buckets - 1) in  
  remove_bucket  h h_buckets i key Bs_internalBuckets.emptyOpt (Bs_Array.unsafe_get h_buckets i)

let rec removeAllBuckets h h_buckets  i key prec buckets =
  match Bs_internalBuckets.toOpt buckets with
  | None -> ()
  | Some {key=k; next} ->
    if  (k : key) = key
    then 
      begin
        (match Bs_internalBuckets.toOpt prec with
         | None -> Bs_Array.unsafe_set h_buckets i  next
         | Some c -> c.next <- next);
        h.size <- h.size - 1;        
      end;
    removeAllBuckets h h_buckets i key buckets next

let removeAll0  h key =
  let h_buckets = h.buckets in 
  (* let i = key_index ~hash h key in *)
  let i = hash key  land (Array.length h_buckets - 1) in  
  removeAllBuckets h h_buckets i key Bs_internalBuckets.emptyOpt (Bs_Array.unsafe_get h_buckets i)


(* TODO: add [removeAll] *)


let rec find_rec  key buckets = 
  match Bs_internalBuckets.toOpt buckets with 
  | None ->
    None
  | Some { key = k; value = d; next =  rest} ->
    if (key : key) = k  then Some d else find_rec key  rest

let findOpt  h (key : key) =
  let h_buckets = h.buckets in 
  (* let nid = key_index ~hash h key in  *)
  let nid = hash key  land (Array.length h_buckets - 1) in 
  match Bs_internalBuckets.toOpt @@ Bs_Array.unsafe_get h_buckets nid with
  | None -> None
  | Some {key = k1; value  = d1; next =  rest1} ->
    if key =k1  then Some d1 else
      match Bs_internalBuckets.toOpt rest1 with
      | None -> None
      | Some {key = k2; value =  d2; next =  rest2} ->
        if  key = k2  then Some d2 else
          match Bs_internalBuckets.toOpt rest2 with
          | None -> None
          | Some { key = k3; value = d3; next =  rest3} ->
            if  key = k3  then Some d3 else find_rec  key rest3


let findAll  h (key : key) =
  let rec find_in_bucket buckets = 
    match Bs_internalBuckets.toOpt buckets with 
    | None ->
      []
    | Some {key = k; value = d; next =  rest} ->
      if  k = key 
      then d :: find_in_bucket rest
      else find_in_bucket rest in
  let h_buckets = h.buckets in     
  (* let nid = key_index ~hash h key in  *)
  let nid = hash key land (Array.length h_buckets - 1) in 
  find_in_bucket (Bs_Array.unsafe_get h_buckets nid)

let rec replace_bucket  (key : key) info buckets = 
  match Bs_internalBuckets.toOpt buckets with 
  | None ->
    true
  | Some ({key = k; value =  i; next} as slot) ->
    if  k = key 
    then
      begin
        slot.key <- key;
        slot.value <- info;
        false
      end
    else
      replace_bucket key info next

let replace  h (key : key) info =
  let h_buckets = h.buckets in 
  (* let i = key_index ~hash h key in *)
  let i = hash key land (Array.length h_buckets - 1) in 
  let l = Array.unsafe_get h_buckets i in  
  if replace_bucket  key info l then begin
    Bs_Array.unsafe_set h_buckets i (Bs_internalBuckets.return {key ; value = info; next = l});
    h.size <- h.size + 1;
    if h.size > Array.length h.buckets lsl 1 then resize  h
  end 

let rec mem_in_bucket (key : key) buckets = 
  match Bs_internalBuckets.toOpt buckets with 
  | None ->
    false
  | Some {key = k; value = d; next =  rest} ->
     k  = key  || mem_in_bucket  key rest     
let mem  h key =
  let h_buckets = h.buckets in 
  (* let nid = (key_index ~hash h key) in  *)
  let nid = hash key land (Array.length h_buckets - 1) in 
  mem_in_bucket  key (Bs_Array.unsafe_get h_buckets nid)


let create = Bs_internalBuckets.create0
let clear = Bs_internalBuckets.clear0
let reset = Bs_internalBuckets.reset0
let length = Bs_internalBuckets.length0
let iter = Bs_internalBuckets.iter0
let fold = Bs_internalBuckets.fold0
let logStats = Bs_internalBuckets.logStats0
let filterMapInplace = Bs_internalBuckets.filterMapInplace0

File: lib/js/bs_HashMapInt.js
'use strict';

var Caml_hash          = require("./caml_hash.js");
var Bs_internalBuckets = require("./bs_internalBuckets.js");

function hash(s) {
  return Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_int(0, s));
}

function insert_bucket(h_buckets, ndata_tail, _, _old_bucket) {
  while(true) {
    var old_bucket = _old_bucket;
    if (old_bucket !== undefined) {
      var key = old_bucket[/* key */0];
      var next = old_bucket[/* next */2];
      var nidx = Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_int(0, key)) & (h_buckets.length - 1 | 0);
      var match = ndata_tail[nidx];
      if (match !== undefined) {
        match[/* next */2] = old_bucket;
      } else {
        h_buckets[nidx] = old_bucket;
      }
      ndata_tail[nidx] = old_bucket;
      _old_bucket = next;
      continue ;
      
    } else {
      return /* () */0;
    }
  };
}

function resize(h) {
  var odata = h[/* buckets */1];
  var osize = odata.length;
  var nsize = (osize << 1);
  if (nsize >= osize) {
    var h_buckets = new Array(nsize);
    var ndata_tail = new Array(nsize);
    h[/* buckets */1] = h_buckets;
    for(var i = 0 ,i_finish = osize - 1 | 0; i <= i_finish; ++i){
      insert_bucket(h_buckets, ndata_tail, h, odata[i]);
    }
    for(var i$1 = 0 ,i_finish$1 = nsize - 1 | 0; i$1 <= i_finish$1; ++i$1){
      var match = ndata_tail[i$1];
      if (match !== undefined) {
        match[/* next */2] = Bs_internalBuckets.emptyOpt;
      }
      
    }
    return /* () */0;
  } else {
    return 0;
  }
}

function add(h, key, value) {
  var h_buckets = h[/* buckets */1];
  var h_buckets_lenth = h_buckets.length;
  var i = Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_int(0, key)) & (h_buckets_lenth - 1 | 0);
  var bucket = /* record */[
    /* key */key,
    /* value */value,
    /* next */h_buckets[i]
  ];
  h_buckets[i] = bucket;
  var h_new_size = h[/* size */0] + 1 | 0;
  h[/* size */0] = h_new_size;
  if (h_new_size > (h_buckets_lenth << 1)) {
    return resize(h);
  } else {
    return 0;
  }
}

function remove_bucket(h, h_buckets, i, key, _prec, _buckets) {
  while(true) {
    var buckets = _buckets;
    var prec = _prec;
    if (buckets !== undefined) {
      var k = buckets[/* key */0];
      var next = buckets[/* next */2];
      if (k === key) {
        if (prec !== undefined) {
          prec[/* next */2] = next;
        } else {
          h_buckets[i] = next;
        }
        h[/* size */0] = h[/* size */0] - 1 | 0;
        return /* () */0;
      } else {
        _buckets = next;
        _prec = buckets;
        continue ;
        
      }
    } else {
      return /* () */0;
    }
  };
}

function remove(h, key) {
  var h_buckets = h[/* buckets */1];
  var i = Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_int(0, key)) & (h_buckets.length - 1 | 0);
  return remove_bucket(h, h_buckets, i, key, Bs_internalBuckets.emptyOpt, h_buckets[i]);
}

function removeAllBuckets(h, h_buckets, i, key, _prec, _buckets) {
  while(true) {
    var buckets = _buckets;
    var prec = _prec;
    if (buckets !== undefined) {
      var k = buckets[/* key */0];
      var next = buckets[/* next */2];
      if (k === key) {
        if (prec !== undefined) {
          prec[/* next */2] = next;
        } else {
          h_buckets[i] = next;
        }
        h[/* size */0] = h[/* size */0] - 1 | 0;
      }
      _buckets = next;
      _prec = buckets;
      continue ;
      
    } else {
      return /* () */0;
    }
  };
}

function removeAll0(h, key) {
  var h_buckets = h[/* buckets */1];
  var i = Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_int(0, key)) & (h_buckets.length - 1 | 0);
  return removeAllBuckets(h, h_buckets, i, key, Bs_internalBuckets.emptyOpt, h_buckets[i]);
}

function find_rec(key, _buckets) {
  while(true) {
    var buckets = _buckets;
    if (buckets !== undefined) {
      var k = buckets[/* key */0];
      var d = buckets[/* value */1];
      var rest = buckets[/* next */2];
      if (key === k) {
        return /* Some */[d];
      } else {
        _buckets = rest;
        continue ;
        
      }
    } else {
      return /* None */0;
    }
  };
}

function findOpt(h, key) {
  var h_buckets = h[/* buckets */1];
  var nid = Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_int(0, key)) & (h_buckets.length - 1 | 0);
  var match = h_buckets[nid];
  if (match !== undefined) {
    var k1 = match[/* key */0];
    var d1 = match[/* value */1];
    var rest1 = match[/* next */2];
    if (key === k1) {
      return /* Some */[d1];
    } else if (rest1 !== undefined) {
      var k2 = rest1[/* key */0];
      var d2 = rest1[/* value */1];
      var rest2 = rest1[/* next */2];
      if (key === k2) {
        return /* Some */[d2];
      } else if (rest2 !== undefined) {
        var k3 = rest2[/* key */0];
        var d3 = rest2[/* value */1];
        var rest3 = rest2[/* next */2];
        if (key === k3) {
          return /* Some */[d3];
        } else {
          return find_rec(key, rest3);
        }
      } else {
        return /* None */0;
      }
    } else {
      return /* None */0;
    }
  } else {
    return /* None */0;
  }
}

function findAll(h, key) {
  var find_in_bucket = function (_buckets) {
    while(true) {
      var buckets = _buckets;
      if (buckets !== undefined) {
        var k = buckets[/* key */0];
        var d = buckets[/* value */1];
        var rest = buckets[/* next */2];
        if (k === key) {
          return /* :: */[
                  d,
                  find_in_bucket(rest)
                ];
        } else {
          _buckets = rest;
          continue ;
          
        }
      } else {
        return /* [] */0;
      }
    };
  };
  var h_buckets = h[/* buckets */1];
  var nid = Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_int(0, key)) & (h_buckets.length - 1 | 0);
  return find_in_bucket(h_buckets[nid]);
}

function replace_bucket(key, info, _buckets) {
  while(true) {
    var buckets = _buckets;
    if (buckets !== undefined) {
      var k = buckets[/* key */0];
      var next = buckets[/* next */2];
      if (k === key) {
        buckets[/* key */0] = key;
        buckets[/* value */1] = info;
        return /* false */0;
      } else {
        _buckets = next;
        continue ;
        
      }
    } else {
      return /* true */1;
    }
  };
}

function replace(h, key, info) {
  var h_buckets = h[/* buckets */1];
  var i = Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_int(0, key)) & (h_buckets.length - 1 | 0);
  var l = h_buckets[i];
  if (replace_bucket(key, info, l)) {
    h_buckets[i] = /* record */[
      /* key */key,
      /* value */info,
      /* next */l
    ];
    h[/* size */0] = h[/* size */0] + 1 | 0;
    if (h[/* size */0] > (h[/* buckets */1].length << 1)) {
      return resize(h);
    } else {
      return 0;
    }
  } else {
    return 0;
  }
}

function mem_in_bucket(key, _buckets) {
  while(true) {
    var buckets = _buckets;
    if (buckets !== undefined) {
      var k = buckets[/* key */0];
      var rest = buckets[/* next */2];
      if (k === key) {
        return /* true */1;
      } else {
        _buckets = rest;
        continue ;
        
      }
    } else {
      return /* false */0;
    }
  };
}

function mem(h, key) {
  var h_buckets = h[/* buckets */1];
  var nid = Caml_hash.caml_hash_final_mix(Caml_hash.caml_hash_mix_int(0, key)) & (h_buckets.length - 1 | 0);
  return mem_in_bucket(key, h_buckets[nid]);
}

var create = Bs_internalBuckets.create0;

var clear = Bs_internalBuckets.clear0;

var reset = Bs_internalBuckets.reset0;

var length = Bs_internalBuckets.length0;

var iter = Bs_internalBuckets.iter0;

var fold = Bs_internalBuckets.fold0;

var logStats = Bs_internalBuckets.logStats0;

var filterMapInplace = Bs_internalBuckets.filterMapInplace0;

exports.hash             = hash;
exports.insert_bucket    = insert_bucket;
exports.resize           = resize;
exports.add              = add;
exports.remove_bucket    = remove_bucket;
exports.remove           = remove;
exports.removeAllBuckets = removeAllBuckets;
exports.removeAll0       = removeAll0;
exports.find_rec         = find_rec;
exports.findOpt          = findOpt;
exports.findAll          = findAll;
exports.replace_bucket   = replace_bucket;
exports.replace          = replace;
exports.mem_in_bucket    = mem_in_bucket;
exports.mem              = mem;
exports.create           = create;
exports.clear            = clear;
exports.reset            = reset;
exports.length           = length;
exports.iter             = iter;
exports.fold             = fold;
exports.logStats         = logStats;
exports.filterMapInplace = filterMapInplace;
/* Bs_internalBuckets Not a pure module */
